packages/clerk-js/src/core/auth/AuthCookieService.ts

```diff
@@ -110,27 +110,47 @@ export class AuthCookieService {
     this.poller.startPollingForSessionToken(() => this.refreshSessionToken());
   }
 
-  private refreshTokenOnVisibilityChange() {
-    document.addEventListener('visibilitychange', () => {
+  private refreshTokenOnFocus() {
+    window.addEventListener('focus', () => {
       if (document.visibilityState === 'visible') {
-        void this.refreshSessionToken();
+        // Certain data-fetching libraries that refetch on focus (such as swr) use setTimeout(cb, 0) to schedule a task on the event loop.
+        // This gives us an opportunity to ensure the session cookie is updated with a fresh token before the fetch occurs, but it needs to
+        // be done with a microtask. Promises schedule microtasks, and so by using `updateCookieImmediately: true`, we ensure that the cookie
+        // is updated as part of the scheduled microtask. Our existing event-based mechanism to update the cookie schedules a task, and so the cookie
+        // is updated too late and not guaranteed to be fresh before the refetch occurs.
+        void this.refreshSessionToken({ updateCookieImmediately: true });
       }
     });
   }
 
-  private async refreshSessionToken(): Promise<void> {
+  private async refreshSessionToken({
+    updateCookieImmediately = false,
+  }: {
+    updateCookieImmediately?: boolean;
+  } = {}): Promise<void> {
     if (!this.clerk.session) {
       return;
     }
 
     try {
-      await this.clerk.session.getToken();
+      const token = await this.clerk.session.getToken();
+      if (updateCookieImmediately) {
+        this.updateSessionCookie(token);
```

Comment:
Again, for any future readers: The reason the token update gets scheduled as a microtask here because it's part of the promise that was initialized when the **async** `refreshSessionToken` was called 

@BRKalow please correct me if I'm wrong here